Fastscape coupling

Djneu opened the pull request:
This is an initial pull request for review and discussion of the code. In the future, I plan to squash some of the commits and split the pull request into a few different pieces to hopefully make it easier to merge and a bit less messy. @anne-glerum @naliboff 

### Before your first pull request:

* [ ] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [ ] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

anne-glerum commented on CMakeLists.txt:
```diff
@@ -477,9 +477,29 @@ IF(NOT ASPECT_ADDITIONAL_CXX_FLAGS STREQUAL "")
   MESSAGE(STATUS "  DEAL_II_CXX_FLAGS_RELEASE: ${DEAL_II_CXX_FLAGS_RELEASE}")
 ENDIF()
 
+# Check if a fastscape library is available, if so link it with aspect.
```
```suggestion
# Check if a FastScape library is available, if so link it with ASPECT.
```

anne-glerum replied:
same below

anne-glerum commented on Named_VTK.f90:
```diff
@@ -0,0 +1,122 @@
+subroutine Fastscape_Named_VTK (f, vex, istep, foldername, k)
+
+    ! subroutine to create a simple VTK file for plotting inside of aspect folder.
```
```suggestion
    ! Subroutine to create a simple VTK file inside of ASPECT's output folder.
```

anne-glerum commented on Named_VTK.f90:
```diff
@@ -0,0 +1,122 @@
+    ! subroutine to create a simple VTK file for plotting inside of aspect folder.
+    ! To use, add this file into the fastscape src directory, and then add
+    ! the line ${FASTSCAPELIB_SRC_DIR}/Named_VTK.f90 to the appropriate place
+    ! in cmakelists.txt
```
Can you make a comment in cmakelists.txt that can be switched on to incorporate this file?

anne-glerum commented on Named_VTK.f90:
```diff
@@ -0,0 +1,122 @@
+    integer i,j
+    character*10 fwtime
+    double precision dx,dy
+
```
rm 1 empty line

anne-glerum commented on Named_VTK.f90:
```diff
@@ -0,0 +1,122 @@
+    dx = xl/(nx - 1)
+    dy = yl/(ny - 1)
+
+    ! Finding the fastscape time incremented by dt. dt may be a double
+    ! that is converted to an integer, but we don't care about then
+    ! exact time that much.
+    ! ftime = (atime + step*dt)
+
+    !Writing the fastscape time to a string, up to 1 billion years. (add another zero if needed eventually)
+   ! write (fwtime,'(i10)') ftime
+   ! if (ftime.lt.10) fwtime(1:9)='000000000'
+   ! if (ftime.lt.100) fwtime(1:8)='00000000'
+   ! if (ftime.lt.1000) fwtime(1:7)='0000000'
+   ! if (ftime.lt.10000) fwtime(1:6)='000000'
+   ! if (ftime.lt.100000) fwtime(1:5)='00000'
+   ! if (ftime.lt.1000000) fwtime(1:4)='0000'
+   ! if (ftime.lt.10000000) fwtime(1:3)='000'
+   ! if (ftime.lt.100000000) fwtime(1:2)='00'
+   ! if (ftime.lt.1000000000) fwtime(1:1)='0'
```
Can these lines be removed?

anne-glerum commented on Named_VTK.f90:
```diff
@@ -0,0 +1,122 @@
+    part1(1:npart1)//'drainage_area'//part2(1:npart2),sngl(a(1:nn)) , &
+    part1(1:npart1)//'catchment'//part2(1:npart2),sngl(catch(1:nn))
+    close(77)
+
```
Can you add a comment that explains what this all does or where it comes from?

anne-glerum commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+  set List of model names = initial temperature
+end
+
+# X boundaries are periodic, and the rest are tangential.
```
This comment does not correspond to what's set

anne-glerum commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+  set Mesh deformation boundary indicators = top : fastscape
+  
+    subsection Fastscape
+      # As the highest resolution at the surface is 4 in the mesh refiniment function, we set this the same.
```
```suggestion
      # As the highest resolution at the surface is 4 in the mesh refinement function, we set this the same.
```

anne-glerum commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+      # As the highest resolution at the surface is 4 in the mesh refiniment function, we set this the same.
+      set Surface resolution = 4
+
+      # Because we only have 1 level of refinement at the surface we set this to 0.
```
```suggestion
      # Because we have the same level of refinement at the surface we set this to 0.
```

gassmoeller replied:
This earlier comment from @anne-glerum is not yet addressed

anne-glerum commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+      # this would be set to 2 (number of cell sizes at surface - 1).
+      set Resolution difference = 0
+
+      # We set fastscape as 1 levels more resolved than ASPECT's initial global surface resolution.
```
```suggestion
      # We set FastScape as 1 level more resolved than ASPECT's initial global surface resolution.
```

anne-glerum commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+      # Seed number for initial topography noise
+      set Fastscape seed = 1000
+
+      # Because no boundaries are periodic, no flux, and no advection we don't need ghost nodes.
```
```suggestion
      # Because no boundaries are periodic, no flux is prescribed, and no advection occurs in FastScape, we do not need ghost nodes.
```

anne-glerum commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+end
+
+
+# We set our global refinement at the resolution level we want at the surface. As of now,
```
What do you mean by `As of now`?

Djneu replied:
I think I thought this could somehow change at some point. I've updated the comment.

anne-glerum commented on include/aspect/geometry_model/box.h:
```diff
@@ -71,6 +71,13 @@ namespace aspect
         Point<dim> get_extents () const;
 
+        /**
+         * Return a point that denotes the repetitions of the box in each dimension
```
```suggestion
         * Return an array that denotes the repetitions of the box's coarse mesh in each direction
```

tjhei replied:
It doesn't return an array?!

anne-glerum commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,269 @@
+     * Define FastScape functions as C functions. Must use exact same function/variable name
+     * and type as used in FastScape. All function names must be made lowercase, and an
+     * underscore added at the end. Types must be defined as pointers, and sent to
+     * fastscape as a reference.
```
```suggestion
     * FastScape as a reference.
```

anne-glerum commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,269 @@
+      void parse_parameters (ParameterHandler &prm);
+
+        /**
+         * A function that fills the viscosity derivatives in the
```
Update comment

anne-glerum reviewed (commented):
part 1

naliboff reviewed (commented):
This is a joint review by @anne-glerum, @Djneu,  and @naliboff.

@Djneu - Thanks for submitting this, really exciting to have this feature in the code! In addition to the specific comments, a few general comments:
1. Place Named_VTK.f90 into a new folder with contrib
2. Can you add a comment in that some of the higher-level reorganization to the mesh deformation sections is included in a separate pull request (and refer to that pull request)
3. Remove changes to unrelated files (artificial_viscosity.h, etc) 
4. Try squashing into fewer commits: 1. major code additions/changes, 2. addition of cookbooks, 3. addition of tests, 4. change log (or squash into one pull request, or make new PR, dealers choice :) )

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+      restart_step = 0;
+
+      // second is for maximum coordinates, first for minimum.
+      for (unsigned int i=0; i<dim; ++i)
```
Note that when the origin is not zero, the maximum coordinates are origin+extents

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+      const unsigned int x_repetitions = geometry->get_repetitions(0);
+      const unsigned int y_repetitions = geometry->get_repetitions(1);
+
+      // Set nx and dx, as these will be the same regardless of dimension.
```
Please explain the logic used here.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+
+      // Sub intervals are 3 less than points, if including the ghost nodes. Otherwise 1 less.
+      table_intervals[0] = nx-1-(2*use_ghost);
+      // TODO: it'd be best to not have to use dim-1 intervals at all.
```
because you always have 2?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+      // Create a folder for the FastScape visualization files.
+      Utilities::create_directory (this->get_output_directory() + "VTK/",
+                                   this->get_mpi_communicator(),
+                                   false);
```
empty line after this

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+      const types::boundary_id relevant_boundary = this->get_geometry_model().translate_symbolic_boundary_name_to_id ("top");
+      const int current_timestep = this->get_timestep_number ();
+
+      double a_dt = this->get_timestep();
```
can it be `const`?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+        }
+
+      // We only want to run fastscape if there was a change in time.
+      if (a_dt > 0)
```
Instead of having a giant if statement, you could also
```
if (this->get_timestep_number() == 0)
  return
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+      if (a_dt > 0)
+        {
+          /*
+           * Initialize a vector of temporary variables to hold: z component, index, Vx, Vy, and Vz.
```
Can you explain why? The fact that a vector of tmp vars is initialized can be seen from the code, why it's necessary not.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+          std::vector<std::vector<double>> temporary_variables(dim+2, std::vector<double>());
+          //Vector to hold the velocities that represent the change to the surface.
+          std::vector<double> V(array_size);
+          //precision = 0.001;
```
rm?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                        if (abs(indx - round(indx)) >= precision)
+                          continue;
+
+                        /*
+                         * If we're in 2D, we want to take the values and apply them to every row of X points.
+                         */
```
```suggestion
                         // If we're in 2D, we want to take the values and apply them to every row of X points.
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                         */
+                        if (dim == 2)
+                          {
+                            for (int ys=0; ys<ny; ys++)
```
```suggestion
                            for (unsigned int ys=0; ys<ny; ys++)
```
?

Djneu replied:
Ny and nx need to be regular int to work with fastscape, so I've left ys as int to avoid warnings.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                                      temporary_variables[i+2].push_back(vel[corner][i]);
+                                  }
+                              }
+                          }
```
```suggestion
                          }
                          // 3D case
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                          }
+                        else
+                          {
+                            //Because indy only gives us the row we're in, we don't need to add 2 for the ghost node.
```
```suggestion
                            // Because indy only gives us the row we're in, we don't need to add 2 for the ghost node.
```
Please clarify.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                          {
+                            //Because indy only gives us the row we're in, we don't need to add 2 for the ghost node.
+                            const double indy = 1+use_ghost+vertex(1)/dy;
+                            if (abs(indy - round(indy)) >= precision)
```
empty line before

anne-glerum replied:
precision is commented out above. Is it set somewhere else?

Djneu replied:
Yes, it's set as a user defined variable now. Though it should really only be useful in the case that one uses fastscape, with use_velocities turned off, and a normal projection free surface.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                      }
+                  }
+
+          // Run fastscape on single processor.
```
Please write `FastScape` everywhere in the comments

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+              const std::string restart_step_filename = dirname + "fastscape_steps_restart.txt";
+              const std::string restart_filename_basement = dirname + "fastscape_b_restart.txt";
+
+	      // Determine whether to create a VTK file this timestep.
```
run `make indent` to fix the indentation

anne-glerum reviewed (commented):
part 2. I'll get back to the rest!

tjhei commented on CMakeLists.txt:
```diff
@@ -477,9 +477,29 @@ IF(NOT ASPECT_ADDITIONAL_CXX_FLAGS STREQUAL "")
+# Check if a fastscape library is available, if so link it with aspect.
+# If the library is not found, then remove fastscape .cc and .h from list
+# so aspect can still compile.
+# TODO: likely need to switch ENV and normal positions so it checks ENV second.
```
I would make the files always compile but fail at runtime if it can not be used. You would introduce a cmake define ASPECT_WITH_FASTSCAPE.

tjhei commented on Named_VTK.f90:
```diff
@@ -0,0 +1,122 @@
+subroutine Fastscape_Named_VTK (f, vex, istep, foldername, k)
+
```
What is this file used for? I don't think it should be on the root directory

Djneu replied:
This is a fastscape file to write visualizations within the specified ASPECT output folder, and show a few more fields than the default fastscape function. I've made a pull request for them to include this within fastscape so hopefully we can remove it from here altogether.

tjhei commented on include/aspect/mesh_deformation/free_surface.h:
```diff
@@ -32,35 +32,6 @@ namespace aspect
   using namespace dealii;
 
-  namespace Assemblers
-  {
```
Can we have this change in a separate PR?

tjhei commented on source/mesh_deformation/interface.cc:
```diff
@@ -38,10 +38,128 @@
+#include <aspect/simulator/assemblers/interface.h>
+#include <aspect/melt.h>
+
+#ifdef ASPECT_USE_PETSC
```
Remove this, please.

tjhei commented on source/postprocess/topography.cc:
```diff
@@ -187,6 +223,13 @@ namespace aspect
                              "Units: years if the "
                              "'Use years in output instead of seconds' parameter is set; "
                              "seconds otherwise.");
+          prm.declare_entry ("Analytical solution of example", "1",
```
This is not a clean solution. Can you make a new plugin that sits on the benchmark folder, instead?

tjhei reviewed (commented):
I am making a few early comments, I did not look at the code on detail, though.

Djneu commented:
Hi all,
Just thought I'd add a bit of info since this pull request is much messier than I thought it would be. There are really only ~5 changes that are needed for the pr, the other changes are mostly from a bad indent that reformatted many files for some reason (e.g., all the files in mesh refinement).

1. Fastscape.cc and fastscape.h
2. Changes to box.cc and box.h to add a few functions
3. Cmakelists.txt
4. Named_VTK.f90, which is a file that needs to be moved to the fastscape source directory on installation
5. fastscape_eroding_box.prm, a new cookbook to test the installation.
6. Changes to free_surface and mesh_deformation/interface that relocates the free surface stabilization factor, though this is open in a separate pull request (#4350).

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                      V2[i-1] = V[index];
+                    }
+                  // Here we use average velocities across the y nodes, excluding the ghost nodes (top and bottom row).
+                  // Note: If ghost nodes are turned off, boundary effects may influence this.
```
Does this require a AssertThrow to make sure users don't turn off the ghost nodes?

Djneu replied:
Hmmm, I'm not sure. I think it is nice to have the option to turn them off. But I'm not sure we've ever really turned them off except for in cases where we were testing something within the coupling.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                  if (slice)
+                    {
+                      const int index = i+nx*(round((ny-1)/2));
+                      V2[i-1] = V[index];
```
In line 650 you convert to m/yr if necessary, here you don't. Please unify.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                        {
+                          idx[0] = j;
+
+                          // Tables are initialized with k = 0 as the bottom and 1 as top, we are only interested in the surface so only fill k==1.
```
Do you ever need the bottom? Can you make a table without it?

Djneu replied:
I may be overlooking something, but the interpolate_boundary_values always causes a crash if I try to only have a top. Otherwise the bottom shouldn't really be needed.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                                                                        table_intervals,
+                                                                        data_table);
+
+          auto lambda = [&](const Point<dim> &p) -> double
```
Can you use a more descriptive name than lambda? See #4444

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                      /*
+                      * Nx*j will give us the row we're in, and one is subtracted as FastScape starts from 1 not zero.
+                      * If we're on the left, the multiple of the row will always represent the first node.
+                      * Subtracting one to the row above this gives us the last node of the previous row.
```
```suggestion
                      * Subtracting one from the row above this gives us the last node of the previous row.
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                      double slope = 0;
+
+                      /*
+                      * Here we set the ghost nodes to the ones next to them, where for the left we
```
```suggestion
                      * Here we set the ghost nodes to the value of the nodes next to them, where for the left we
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                      if (current_timestep == 1 || left_flux == 0)
+                        {
+                          /*
+                           * If its the first timestep add in initial slope. If we have no flux,
```
```suggestion
                           * If it's the first timestep add in initial slope. If we have no flux,
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                          /*
+                          * If nodes on both sides are going the same direction, then set the respective
+                          * ghost nodes to equal these sides. By doing this, the ghost nodes at the opposite
+                          * side of flow will work as a mirror mimicing what is happening on the other side.
```
```suggestion
                          * side of flow will work as a mirror mimicking what is happening on the other side.
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                          * ghost nodes to equal these sides. By doing this, the ghost nodes at the opposite
+                          * side of flow will work as a mirror mimicing what is happening on the other side.
+                          */
+                          if (vx[index_right-1] > 0 && vx[index_left+1] >= 0)
```
Why can one side be >= instead of >?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                      vz[index_top] = vz[index_top-nx];
+
+                      vy[index_bot] = vy[index_bot+nx];
+                      vy[index_top] =  vy[index_top-nx];
```
```suggestion
                      vy[index_top] = vy[index_top-nx];
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                      vy[index_top] =  vy[index_top-nx];
+
+                      vx[index_bot] = vx[index_bot+nx];
+                      vx[index_top] =  vx[index_top-nx];
```
```suggestion
                      vx[index_top] = vx[index_top-nx];
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                            Patterns::Double(0),
+                            "Maximum timestep for FastScape.");
+          prm.declare_entry("Vertical exaggeration", "-1",
+                            Patterns::Double(),
```
or ints?

Djneu replied:
For the vertical exaggeration, FastScape expects a double value.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                             "average the inner third of what FastScape calculates.");
+          prm.declare_entry("Fastscape seed", "1000",
+                            Patterns::Integer(),
+                            "Seed used for adding an initial 0-20 m noise to FastScape topography.");
```
I think I saw 5 m mentioned in the code?

Djneu replied:
Yes, and this is a user set parameter now. I've updated the description here and the comment in the code.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+          prm.declare_entry ("Use marine component", "false",
+                             Patterns::Bool (),
+                             "Flag to use the marine component of FastScape.");
+          prm.declare_entry ("Use stratigraphy", "false",
```
Should we turn this off? We never really tested it and expect problems right?

Djneu replied:
Yes, I think we decided to completely remove the stratigraphy component.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                            "Total number of steps you expect in the FastScape model, only used if stratigraphy is turned on.");
+          prm.declare_entry("Number of horizons", "1",
+                            Patterns::Integer(),
+                            "Number of horizons to track and visualize in FastScape.");
```
Only used if stratigraphy is turned on?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                               "Left boundary condition, where 1 is fixed and 0 is reflective.");
+            prm.declare_entry("Left mass flux", "0",
+                              Patterns::Double(),
+                              "Flux per unit length through left boundary (m^2/yr)");
```
Please use the correct formatting of units here and in all `declare_entry`'s, e.g. `Units: $\\si{m^2/s}$.`.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+      end_time = prm.get_double ("End time");
+      if (prm.get_bool ("Use years in output instead of seconds") == true)
+        end_time *= year_in_seconds;
+      prm.enter_subsection ("Mesh deformation");
```
empty line above

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+
+              if (sr_values.size() != sr_times.size()+1)
+                  AssertThrow(false, ExcMessage("Error: There must be one more sediment rain value than interval."));
+
```
rm empty line

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+  {
+    ASPECT_REGISTER_MESH_DEFORMATION_MODEL(FastScape,
+                                           "fastscape",
+                                           "A plugin which uses the program FastScape to add surface processes "
```
```suggestion
                                           "A plugin that uses the program FastScape to add surface processes "
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                                           "such as diffusion, sedimentation, and the Stream Power Law to ASPECT. "
+                                           "FastScape is initialized with ASPECT height and velocities, and then "
+                                           "continues to run in the background, updating with new ASPECT velocities "
+                                           "when called. Once FastScape has run for the given amount of timesteps, it "
```
```suggestion
                                           "when called. Once FastScape has run for the given amount of timesteps per ASPECT timestep, it "
```

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+                                           "FastScape is initialized with ASPECT height and velocities, and then "
+                                           "continues to run in the background, updating with new ASPECT velocities "
+                                           "when called. Once FastScape has run for the given amount of timesteps, it "
+                                           "compares the initial and final heights to send a velocity back to the mesh "
+                                           "deformation plugin.")
```
```suggestion
                                           "this plugin compares the initial and final heights to compute a vertical mesh deformation velocity at the surface boundary. ")
```

anne-glerum reviewed (commented):
Hi Derek, I finished my first general review of the code. There are many small things that need to be addressed before I can take a more detailed look at the way things are implemented. Very much looking forward to seeing this merged!

anne-glerum commented:
Hi Derek, I know you were thinking of opening up a new PR to get rid of the files that were included by accident here (as you explain in your last comment), but I finished my review here to keep all the comments on the same files in the same place. Also, there might be a solution where we can keep one PR? I can think of checking out the old versions of the specific files and committing them. By squashing at least the edits and their undoing into one commit, maybe the changes will disappear. But maybe that's more hassle than it's worth. @tjhei would you have a solution?

tjhei commented:
> would you have a solution?

1. delete all files that are not necessary with ``git rm``, commit
2. run ``make indent``, commit
3. do a ``git rebase -i`` and squash all your commits into one
4. force push

anne-glerum commented:
For documenting to address in the future:
Running a model that has an origin that is not at (0,0) leads to a crash. I have not investigated it further, but it will fail after writing the Initial VTK file in t1.
Error message:
```
*** Error in `/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther': free(): invalid size: 0x0000000005bd4070 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x81329)[0x2aaad167b329]
/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther(_ZNK6aspect15MeshDeformation9FastScapeILi2EE40compute_velocity_constraints_on_boundaryERKN6dealii10DoFHandlerILi2ELi2EEERNS3_17AffineConstraintsIdEERKSt3setIjSt4lessIjESaIjEE+0x2a65)[0x1387045]
/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther(_ZN6aspect15MeshDeformation22MeshDeformationHandlerILi2EE16make_constraintsEv+0x7fd)[0x13a02cd]
/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther(_ZN6aspect15MeshDeformation22MeshDeformationHandlerILi2EE7executeEv+0xc0)[0x13b1ab0]
/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther(_ZN6aspect9SimulatorILi2EE14solve_timestepEv+0x87)[0xa86f57]
/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther(_ZN6aspect9SimulatorILi2EE3runEv+0x2d7)[0xaa9467]
/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther(_Z13run_simulatorILi2EEvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES7_bbb+0x279)[0x1572f89]
/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther(main+0x715)[0x9f54b5]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x2aaad161c555]
/home/projects/bbp00039/fastscape_build_files/aspect-22feb22-dealii-10.0-pre-fastscape-esther[0x9fc9b7] 
```

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+
+              /*
+               * The ghost nodes are added as a single layer of nodes surrounding the entire model,
+               * so that 3x3 amount of nodes representing ASPECT will become 5x5 in FastScape, with the inner
+               * 3x3 being returned to ASPECT.
```
I tried to word this better now, I meant to say that if we assume we have an ASPECT surface that is 3 nx points, and 3 ny points, the FastScape surface with ghost nodes will then be a 2D mesh that is 5 nx and 5 ny points, but on returning the surface to ASPECT we don't consider the outer points, so only the inner 3 nx and 3 ny are seen by ASPECT. Maybe it makes more sense to not use a specific example and instead say if we assume ASPECT is an nx*ny surface mesh, then FastScape will be (nx+2)*(ny+2), but will only return the inner nx*ny that corresponds to ASPECT?

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+
+              fastscape_set_h_(h.get());
+
+              // The visualization is 1 step behind the current timestep.
```
I set it this way because on the first timestep, we create an initial VTK file before running FastScape and a second one after running it. Though, the visualization step only changes the file name on the VTK files.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+          data_table.TableBase<dim,double>::reinit(size_idx);
+          TableIndices<dim> idx;
+
+          // Loop through the data table and fill out the surface (i or k = 1) to the velocities from FastScape.
```
I think I was referring to k = 1 being the surface, but this doesn't matter as after your suggestion we now fill both k=0 and k=1 with the values (at least after the next commit!).

Djneu commented:
Hi all, 
I've hopefully addressed all the previous comments and fixed the indent issue. I think the plugin is now in a place for a second round of reviews. I'll note, there is a Named_VTK.f90 file that I have not updated and is still a bit of a mess, I'm hoping this will eventually be included within the fastscape code so I've left it alone for now. @gassmoeller @tjhei @naliboff @anne-glerum

gassmoeller commented on CMakeLists.txt:
```diff
@@ -479,9 +479,11 @@ IF(NOT ASPECT_ADDITIONAL_CXX_FLAGS STREQUAL "")
 ADD_EXECUTABLE(${TARGET} ${TARGET_SRC})
 DEAL_II_SETUP_TARGET(${TARGET})
 
+
```
Please remove these lines again, unless you added them on purpose (same with the line above).

gassmoeller commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+# At the top, we define the number of space dimensions we would like to
```
We changed the structure of our cookbooks a bit. Could you create a new directory `cookbooks/fastscape_eroding_box` and put the .prm file inside? Also can you either: 1. Link to a paper where we can read more about this model, or 2. Add a subdirectory `doc` inside the cookbook folder and write a paragraph (or more) about what is happening inside this model?

Djneu replied:
This prm file is only a quick test of an eroding central block to make sure the coupling is working after installation or when making changes to the plugin. Is there a better place to put it?

gassmoeller replied:
That sounds like a test. However to make this into an official test we would need to integrate downloading and building fastscape into the test suite. Can you just put it into its own subdirectory at least? Maybe add what you just wrote into either a README.md, or a very small file in cookbooks/fastscape_eroding_box/docs.

gassmoeller commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+
+set Use years in output instead of seconds = true
+set End time                               = 2.5e5
+set Maximum time step                      = 50000 #5e6
```
Please remove all commented values from the file (unless they serve a purpose, then add a sentence on what the commented value means).

gassmoeller commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+        set Left   = 1
+      end
+
+      # We use both SPL and diffusion. The bedrock deposition
```
What is SPL? please write out

gassmoeller commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+  set Model name = vertical
+
+  subsection Vertical
+    set Magnitude = 0   # = Ra
```
```suggestion
    set Magnitude = 0
```

gassmoeller commented on cookbooks/fastscape_eroding_box.prm:
```diff
@@ -0,0 +1,172 @@
+end
+
+
+# We set our global refinement at the resolution level we want at the surface. As of now,
```
```suggestion
# We set our global refinement to the refinement level we want at the surface. As of now,
```

gassmoeller commented on include/aspect/simulator_access.h:
```diff
@@ -142,27 +142,32 @@ namespace aspect
-   *
-   * @ingroup Simulator
-   */
+  * SimulatorAccess is a base class for different plugins like postprocessors.
+  * It provides access to the various variables of the main class that
+  * plugins may want to use in their evaluations, such as solution vectors,
+  * the current time, time step sizes, material models, or the triangulations
+  * and DoFHandlers that correspond to solutions.
+  *
+  * This class is the interface between plugins and the main simulator class.
+  * Using this insulation layer, the plugins need not know anything about the
+  * internal details of the simulation class.
+  *
+  * Every Postprocessor is required to derive from SimulatorAccess. It is
+  * optional for other plugins like MaterialModel, GravityModel, etc..
+  *
+  * Since the functions providing access to details of the simulator class
+  * are meant to be used only by derived classes of this class (rather than
+  * becoming part of the public interface of these classes), the functions of
+  * this class are made @p protected.
+  *
+  * @ingroup Simulator
+  */
```
You indented all of these lines slightly differently. Please restore the original indentation (one more space per line).

gassmoeller commented on include/aspect/simulator_access.h:
```diff
@@ -280,6 +285,14 @@ namespace aspect
 
+      /**
+       * Return a reference to the manager of the time stepping strategies.
+       * This can then, for example, be check whether a checkpoint needs to
```
```suggestion
       * This can then be used, for example, to check whether a checkpoint needs to
```

gassmoeller commented on include/aspect/geometry_model/box.h:
```diff
@@ -71,6 +71,13 @@ namespace aspect
+         * Return an integer array that denotes the number of repetitions of
+         * the box's coarse mesh.
+         */
+        std::array<unsigned int,dim>
+        get_repetitions () const;
```
If you write this as:
```suggestion
        const std::vector<unsigned int> &
        get_repetitions () const;
```
You can make the function much easier (see my comment below). (also update the comment above when you do this).

bangerth replied:
This comment still applies.

gassmoeller commented on source/geometry_model/box.cc:
```diff
@@ -228,6 +228,19 @@ namespace aspect
+    std::array<unsigned int,dim>
+    Box<dim>::get_repetitions () const
+    {
+      std::array<unsigned int,dim> rep;
+      for (unsigned int i=0; i<dim; ++i)
+        {
+          rep[i] = repetitions[i];
+        }
+
+      return rep;
```
If you change the declaration of this function as suggested above, you can do this instead:
```suggestion
      return repetitions;
```

gassmoeller replied:
Ah, I did not see that `repetitions` is of type `unsigned int [3]`, which is awkward to use. Keep your version for now, I will make a separate PR to change it later.

bangerth replied:
If you rebase, you'll get @gassmoeller's changes made in #5098 that make it possible what he suggested.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -479,6 +479,20 @@ namespace aspect
 
+        // These are the compositional fields used to track deposition depth and
+        // sediment age of the sediments deposited by the FastScape plugin.
+        // ToDo: insert check that these fields are only set as a mask when fastscape is used?
```
Just convention:
```suggestion
        // TODO: insert check that these fields are only set as a mask when fastscape is used?
```
And I would remove this TODO and check. Someone could create a field named "sediment_age" without intending to use Fastscape, and it would come as a surprise if the model fails then.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -479,6 +479,20 @@ namespace aspect
+        // ToDo: insert check that these fields are only set as a mask when fastscape is used?
+        if (this->introspection().compositional_name_exists("sediment_age"))
+          {
+            const int sedi_age_position_tmp = this->introspection().compositional_index_for_name("sediment_age");
```
```suggestion
            const unsigned int sedi_age_position_tmp = this->introspection().compositional_index_for_name("sediment_age");
```

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -479,6 +479,20 @@ namespace aspect
+          }
+        if (this->introspection().compositional_name_exists("deposition_depth"))
+          {
+            const int depo_depth_position_tmp = this->introspection().compositional_index_for_name("deposition_depth");
```
```suggestion
            const unsigned int depo_depth_position_tmp = this->introspection().compositional_index_for_name("deposition_depth");
```

gassmoeller commented on CMakeLists.txt:
```diff
@@ -525,6 +527,22 @@ IF(ASPECT_WITH_LIBDAP)
+IF(ASPECT_WITH_FASTSCAPE)
+  IF (EXISTS ${FASTSCAPE_DIR}/libfastscapelib_fortran.so)
+     MESSAGE(STATUS "FastScape library found at ${FASTSCAPE_DIR}/libfastscapelib_fortran.so")
+     FIND_LIBRARY(FASTSCAPE NAMES fastscapelib_fortran PATHS $ENV{FASTSCAPE_DIR} ${FASTSCAPE_DIR} PATH_SUFFIXES lib NO_DEFAULT_PATH)
```
Should we include a minimum version here, like we do for deal.II above? Just to make sure no one tries to build this with very old Fastscape versions that may not be compatible.

gassmoeller commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,680 @@
+    void fastscape_copy_f_(double *sf);
+
+    /**
+     * Initialize the straigraphy component of FastScape.
```
```suggestion
     * Initialize the stratigraphy component of FastScape.
```

gassmoeller commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,680 @@
+  /**
+   * A plugin that utilizes the landscape evolution code FastScape
+   * to deform the ASPECT boundary through advection, uplift,
+   * hillslope diffusion, the stream power law, sediment deposition
```
"the stream power law" is a bit opaque for users not familiar with surface evolution. Is this the part that describes river incision? Maybe reformulate?

Djneu replied:
Yes, the stream power law is for river incision, I added this to the description.

gassmoeller commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,680 @@
+       * Function used to set the FastScape ghost nodes. FastScape boundaries are
+       * not uplifted or periodic for advection and diffusion. By using a layer
+       * of extra nodes in the FastScape model, we can avoid seeing FastScape boundary
+       * effects within ASPECT. Similalrly, we can use these nodes to have fully
```
```suggestion
       * effects within ASPECT. Similarly, we can use these nodes to have fully
```

gassmoeller commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,680 @@
+       * ghost nodes, and the nodes one layer inward (boundary nodes in ASPECT) to match
+       * the parameters on the other side (vx, vy, vz, h). This is done every ASPECT timestep
+       * before running FastScape.
+       * TODO: Should this be in private?
```
Yes, if this function is never called from outside this class, it should be private. It is good practice to hide all unnecessary functions from outside access (because it makes it easier for you to figure out later, where this function could be called from, and reduces the potential for tight coupling between different  code pieces).

Djneu replied:
Good to know! In that case I've moved all the new functions into private.

gassmoeller commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,680 @@
+      void execute_fastscape(double *h, double *kd, int istep) const;
+
+      /**
+       * Function to apply orographic controls to the FastScape model.
```
Again as someone not familiar with surface evolution, can you define "orographic" here?

gassmoeller reviewed (commented):
First batch of comments on the interface and everything but fastscape.cc.

First of all thanks for the massive amount of work, this is very exciting and I am looking forward to seeing this merged! The code looks good in many ways already, I think the main challenge will be to find a compromise between our naming and coding conventions and Fastscape's conventions (e.g. in terms of variable names and the types of parameters to be handed over to the C functions). I will think about this a bit and comment on it when I have read through the .cc file. All my comments so far should be unaffected by that. Also please go through the older comments by @anne-glerum, I have seen several that were not yet addressed.

gassmoeller commented:
Also you should definitely add a changelog entry for this, and give us a reference to any publications that should be cited if users enable the fastscape coupling (we will add them here: https://aspect.geodynamics.org/citing.html?ver=release&src=www)

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+          // Find  timestep size, run fastscape, and make visualizations.
+          execute_fastscape(h.get(), kd.get(), istep);
+
+          // Write a file to store h, b & step for restarting.
+          // TODO: It would be good to roll this into the general ASPECT checkpointing,
+          // and when we do this needs to be changed.
+          if (((this->get_parameters().checkpoint_time_secs == 0) &&
+               (this->get_parameters().checkpoint_steps > 0) &&
+               ((current_timestep + 1) % this->get_parameters().checkpoint_steps == 0)) ||
+              (this->get_time() + a_dt >= end_time && this->get_timestepping_manager().need_checkpoint_on_terminate()))
+            {
+              save_restart_files(h.get(), b.get(), sf.get(), istep);
+            }
```
There are two ways to connect this into the general checkpointing mechanism:
1. Create a checkpointing architecture with `serialize` function like we have for the postprocessors (see, e.g. here https://github.com/geodynamics/aspect/blob/18d75ebaaab2d409685b87633701de676c00720f/source/postprocess/visualization.cc#L314). This would allow you to store all the data as part of ASPECT's main checkpoint files. However, this architecture does not exist yet for mesh deformation plugins and would need a bit of work.
2. Make use of the `pre_checkpoint_store_user_data` signal that is triggered every time before a checkpoint is written. You can see how this signal is used here: https://github.com/geodynamics/aspect/blob/18d75ebaaab2d409685b87633701de676c00720f/source/particle/world.cc#L218. I think this is the easier approach. To make this work, you would move all commands that need to be executed on checkpointing into a separate function (`prepare_checkpoint`) and then connect this function to the signal in the `initialize` function like this:
```
this->get_signals().pre_checkpoint_store_user_data.connect(
            [&] (typename parallel::distributed::Triangulation<dim> &)
          {
            prepare_checkpoint();
          });
```
 
Once connected the `prepare_checkpoint` function will be executed every time a checkpoint is written.


Similarly you could use the `post_resume_load_user_data` signal to load data after restarting from a checkpoint.

Djneu replied:
I tried to use the pre_checkpoint_store_user_data, but I seem to be running into a problem where by the time the signal is called, the h data that exists only on processor 0 during the compute_velocity_constraints_on_boundary is no longer accessible. I think a way around this could be to add 3 mutable vectors to fastscape.h, and then resize  these to array_size during initialize and update them with the relevant data while still on processor zero within the function. Does this seem like an okay way to do this, or am I missing an easier way to do it?

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+            }
+
+          // If we've reached the end time, destroy FastScape.
+          if (this->get_time() + this->get_timestep() > end_time)
```
This will only work for the most common termination criteria (end time), but not for things like end step or steady state velocity. Can you move this into the destructor of this class? I am not sure using `this->pcout()` in the destructor is safe, but it is worth a try. Otherwise we could look into more complicated solutions, like asking the termination manager here, if the simulation will be executed after this time step. But this could have unintended consequences, as this operation changes the state of some termination criteria like the steady state computation. Let's try the destructor first. Do you know how to move this?

Djneu replied:
I've moved it into MeshDeformationHandler<dim>::~MeshDeformationHandler () in interface.cc, is this correct? pcout did work here, but it probably isn't necessary so I've removed it.

gassmoeller replied:
I meant to create a new destructor `Fastscape::~Fastscape`. That would be the best place to destroy any objects the Fastscape class has created.

gassmoeller commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,680 @@
+       * Applied as either a + or - value to the topography
+       * such that the total difference can be up to 2*noise_h.
+       */
+      int noise_h;
```
Why is this an `int` and not a `double`?

gassmoeller commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,680 @@
+      /**
+       * Size of the FastScape array (nx*ny).
+       */
+      int array_size;
```
can this be an `unsigned int`? That would be more precise. Same question for nx and ny above.

Djneu replied:
I thought I had checked this previously, but array_size isn't called by fastscape so this can be an unsigned int. Ny and nx, however, cannot be unsigned. I've updated all the array_size loops to reflect this.

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+
+          // Find out our velocities from the change in height.
+          // Where V is a vector of array size that exists on all processes.
+          for (int i=0; i<array_size; i++)
```
Here and in other loops: Apart from very specific cases we always loop like this (only change the `unsigned int` if you can make `array_size` an unsigned int):
```suggestion
          for (unsigned int i=0; i<array_size; ++i)
```

There is no difference for simple types like `int`, but it is useful to keep to one style and `++i` has advantages for complex types.

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+
+      // Get a quadrature rule that exists only on the corners, and increase the refinement if specified.
+      const QIterated<dim-1> face_corners (QTrapez<1>(),
+                                           pow(2,additional_refinement+surface_refinement_difference));
```
```suggestion
                                           std::pow(2,additional_refinement+surface_refinement_difference));
```

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+                                        update_values |
+                                        update_quadrature_points);
+
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = this->get_dof_handler().begin_active(),
+      endc = this->get_dof_handler().end();
+
+      for (; cell != endc; ++cell)
```
There is a simpler way to write this by now:
```suggestion
for (const auto &cell : this->get_dof_handler().active_cell_iterators())
```

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+                    // Find what x point we're at. Add 1 or 2 depending on if ghost nodes are used.
+                    const double indx = 1+use_ghost_nodes+(vertex(0) - grid_extent[0].first)/dx;
+
+                    // If our x or y index isn't close to a whole number, then it's likely an artifact
+                    // from using an over-resolved quadrature rule, in that case ignore it.
+                    if (abs(indx - round(indx)) >= precision)
+                      continue;
```
Can you elaborate a bit on this? Are you saying the accuracy of the quadrature point position would be too low? That seems unlikely. Or are you saying someone could choose a different refinement level for the quadrature rule than the Fastscape mesh? This would look like an error to me that should not happen, so we should assert it does not happen (not ignore the false points).

Also if we ignore these points, wouldnt we have gaps in the data arrays we are filling here?

Djneu replied:
This was done for two reasons:

1. Having multiple resolutions at the surface. Right now, a single quadrature rule exists for all the cells that makes sure to make enough interpolation points that the lowest resolution expected at the surface has enough points to fill the fastscape mesh. However, because this same rule is used on higher resolution cells, there will be extra points that do not correspond to the fastscape mesh, and this should skip over them. Is there perhaps a more straight forward way to do this?

3. The precision variable is more for if fastscape is used with a normal projection free surface, in which case the surface points likely won't match up exactly with the fixed fastscape mesh, and the quickest solution was to find a point that was near enough. Though maybe it would be better to not use the normal projection with fastscape and avoid this altogether.

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+                        // Because indy only gives us the row we're in, we don't need to add 2 for the ghost node.
+                        const double indy = 1+use_ghost_nodes+(vertex(1) - grid_extent[1].first)/dy;
+
+                        if (abs(indy - round(indy)) >= precision)
+                          continue;
```
same question as above

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+
+
+    template <int dim>
+    void FastScape<dim>::fill_fastscape_arrays(double *h, double *kd, double *kf, double *vx, double *vy, double *vz, std::vector<std::vector<double>> temporary_variables) const
```
Please put each parameter on its own line (same for all other functions in the .cc and .h file). Makes it easier to read.

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+    template <int dim>
+    void FastScape<dim>::execute_fastscape(double *h, double *kd, int istep) const
+    {
+      const double a_dt = this->get_timestep()/year_in_seconds;
```
```suggestion
      const double timestep_in_years = this->get_timestep() / year_in_seconds;
```

gassmoeller reviewed (commented):
Second batch of comments. I will need one more before I am through, this is a pretty significant amount of code. This should give you something to work with already.

gassmoeller reviewed (commented):
More comments about `execute_fastscape`

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+          else
+            {
+              // If it isn't the first timestep we ignore initialization and instead copy all height values from FastScape.
+              if (use_velocities)
```
This was added in case someone wanted to use the free surface in combination with fastscape, in which case fastscape won't advect/uplift the mesh. Generally, this is set to true and we copy the likely higher resolution surface topography directly from fastscape. Otherwise, if the free surface is on, we wanted to use the updated/advected surface we gathered from ASPECT.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+            // Outputs new h values
+            fastscape_copy_h_(h);
+          }
+        while (istep<fastscape_iterations);
```
Thanks for the suggestion! I had done this very early on and never looked at it again for some reason, you are correct there is no need to do anything but execute the step within the loop. Also, I realized istep is no longer used for anything in the plugin and I have removed it.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+      for (unsigned int i=0; i<dim; ++i)
+        {
+          grid_extent[i].first = geometry->get_origin()[i];
+          grid_extent[i].second = geometry->get_extents()[i];
```
Yes, the second contains the extent and not the maximum coordinates. I've updated the comment.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+      fastscape_iterations = fastscape_iterations + istep;
+      this->get_pcout() << "   Calling FastScape... " << (fastscape_iterations-istep) << " timesteps of " << f_dt << " years." << std::endl;
+      {
+        auto t_start = std::chrono::high_resolution_clock::now();
```
I seem to get an infinite hang when doing it this way, however it works when only using one processor. There is a second timer in the plugin that is done this way and works, but the bracket scope doesn't exist only on one processor. Any ideas how to get around this?

gassmoeller replied:
Yes, if you have an else branch to the part that is only executed on one processor: enter the same TimerOutput::Scope there as well. When leaving a timer scope all processors exchange information about their timings to form an aggregated timing (e.g. average), so it is always necessary that all processors enter and leave scopes (even if they follow different code paths).

Djneu commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,269 @@
+        double output_interval;
+
+        /**
+         * A time (in seconds) at which the last graphical output was supposed
```
We compare this to this->get_time(), end_time, and output_interval. I think this->get_time() is always in seconds, and for the other two we convert them to seconds in the case that years are specified. Does this work?

gassmoeller reviewed (commented):
Some replies to your questions and a few more comments I had made a while ago, but somehow never made it into a review.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1632 @@
+            {
+              // If it's the first timestep add in initial slope. If we have no flux,
+              // set the ghost node to the node next to it.
+              // FastScape calculates the slope by looking at all nodes surrounding the point
+              // so we need to consider the slope over 2 dx.
```
Yes, something like this. I had another look at their code, and it looks like the slope for e.g., h[1,1]  in x is actually calculated like: ((h[2,2] + 2*h[2,1] + h[2,0]) - (h[0,2] + 2*h[0,1] + h[0,0])/(8*dx). For this we can only adjust h[0,i]  to define the slope at h[1,i]. I think I decided using 2*dx when looking at a single point was close enough as we go through all of them in order, but it seems like there is probably a better way to do this.

Djneu commented:
Hi all, thank you for all the suggestions! Other than a few open questions I think I've addressed all the previous comments and the code is ready for another review. @anne-glerum @naliboff @gassmoeller @tjhei @bangerth

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1701 @@
+          kf[i] = kff;
+          kd[i] = kdd;
+
+          if (h[i] == std::numeric_limits<double>::max())
```
Reminder for me to switch this to look at vx instead of h. As of now, it should only work for the first timestep and not work if the mesh refinement is updated at the surface.

gassmoeller reviewed (commented):
Very nice, getting close. I have some remaining comments, mostly small code style issues.

The only bigger issue is how difficult it is to integrate Named_VTK into Fastscape and that ASPECT doesnt compile without doing that. Can you either:
- find a way to make that simpler, e.g. a script that can copy the file and patch the CMakeLists of Fastscape as necessary
- include a patched version of fastscape in the contrib folder (like we do for `world_builder`)
- add another cmake variable that disables the Fastscape VTK output by default so that we can compile with a standard fastscape?

gassmoeller commented:
Oh, you also mentioned open questions in your last post. Can you summarize which questions are open? Most previous comments are marked as resolved or outdated, so it is a bit hard to keep track of what is resolved and what is still open.

tjhei commented:
Instead of us bundling a patched fastscape or providing these complicated instructions, can you maybe fork fastscape into your own github repo and apply your patches?

Djneu commented on contrib/fastscape/Named_VTK.f90:
Hi both, sorry that was an outdated named_vtk file, in the new one big_endian is no longer specified. However, this seems to always run into an issue where fastscape works fine but the visualizations do not show up. The fix I've been using for this is to add 

export GFORTRAN_CONVERT_UNIT='big_endian'

to the .bashrc (https://github.com/fastscape-lem/fastscapelib-fortran/issues/49). Anne, I tried your suggested CMakeLists change but with this I was still running into the visualization issue when not exporting this. Do the visualizations work for you without having to export the convert unit?

bangerth replied:
Do you happen to know where libfastscape.so checks for this environment variable? Does it do that once, upon startup of the library, or does it do it directly before writing a VTK file? If the latter, we could call `std::setenv` right before calling the fastscape function. If the former, one *could* presumably work around this somehow (using `dlopen()`, and calling `setenv()` just before) but that's a bit more work.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1701 @@
+                             Patterns::Bool (),
+                             "If this is set to true, then a 2D model will only consider the "
+                             "center slice FastScape gives. If set to false, then aspect will"
+                             "average the inner third of what FastScape calculates.");
```
Yes, this was how we did it a long while back, you are correct now it should average everything excluding the ghost nodes.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1701 @@
+          interpolation_extent[i].second = (grid_extent[i].second + grid_extent[i].first);
+        }
+
+      Functions::InterpolatedUniformGridData<dim> *velocities;
+      velocities = new Functions::InterpolatedUniformGridData<dim> (interpolation_extent,
+                                                                    table_intervals,
+                                                                    velocity_table);
```
Yes, this worked. Thanks!

Djneu commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,721 @@
+       * the parameters on the other side (vx, vy, vz, h). This is done every ASPECT timestep
+       * before running FastScape.
+       */
+      void set_ghost_nodes(double *h,
```
Okay, I've updated all the functions. Thank you for the explanation as well!

Djneu commented:
Hi all, thanks for all the suggestions! I've tried to address most of the new comments and updated the ghost nodes quite a bit. For the named_vtk file, I've removed this and done as @tjhei suggested and made the changes in the default branch here: https://github.com/Djneu/fastscapelib-fortran
Also, note that at the moment "export GFORTRAN_CONVERT_UNIT='big_endian'" must be used for the VTK files to work properly.

There are a few open questions, the more major ones being:

1. Whether there is a better way to do the quadrature rule for having enough interpolation points to fill the fixed fastscape resolution mesh. https://github.com/geodynamics/aspect/pull/4396#discussion_r1121663580

2. Somewhat related to the last question, at the moment having precision be a user-defined variable and the variable "use_velocities" was a work around to be able to use fastscape with the free surface. This comes with its own issues with the mesh not being fixed using normal projection, and losing topography resolution when interpolating from a lower resolution ASPECT mesh onto a higher resolution FastScape one. Generally, we've always let fastscape handle advection, in which case should we keep this option available and possibly try to improve it at some point or remove it?

3. Right now the restart files are still written outside the general ASPECT checkpointing, and I was having trouble getting the pre_checkpoint function to work. https://github.com/geodynamics/aspect/pull/4396#discussion_r1124365806

Some minor questions that could be changed but are maybe not necessary:
    https://github.com/geodynamics/aspect/pull/4396#discussion_r1157104612  - mass flux. (Rene)
    https://github.com/geodynamics/aspect/pull/4396#discussion_r905067660   - boundary condition integers (John)
    https://github.com/geodynamics/aspect/pull/4396#discussion_r1124375375  - get time.  (Anne)
    https://github.com/geodynamics/aspect/pull/4396#discussion_r1081231170  - years vs. seconds (Anne)
    https://github.com/geodynamics/aspect/pull/4396#discussion_r774690022   - assert throw for ghost nodes. (Anne)
    https://github.com/geodynamics/aspect/pull/4396#discussion_r774692317   - bottom row for table. (Anne)
    https://github.com/geodynamics/aspect/pull/4396#discussion_r774695522   - ghost node periodicity (Anne)
    
   A final thing to note, the written FastScape VTK's do not include a time. There is a separate script we generally run after the model finishes to convert these from VTK to a VTS file that includes the correct timing to match ASPECT. I can include this later, but maybe this conversion is something we could also have ASPECT do automatically after FastScape writes the file?

bangerth commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -479,6 +479,18 @@ namespace aspect
               composition_mask.set(i,false);
           }
 
+        // Mask fields that track the age and deposition depth of deposited sediments.
+        if (this->introspection().compositional_name_exists("sediment_age"))
+          {
+            const unsigned int sedi_age_position_tmp = this->introspection().compositional_index_for_name("sediment_age");
+            composition_mask.set(sedi_age_position_tmp,false);
+          }
+        if (this->introspection().compositional_name_exists("deposition_depth"))
+          {
+            const unsigned int depo_depth_position_tmp = this->introspection().compositional_index_for_name("deposition_depth");
+            composition_mask.set(depo_depth_position_tmp,false);
+          }
```
This is unfortunate that you need to mask these field names in individual material models. Could this whole block be guarded somehow with an `if` that checks whether you use FastScape or not?

bangerth reviewed (commented):
Here's a bunch of (mostly small) comments about everything but the `fastscape.h` and `fastscape.cc` files. I'll see that I get to those tomorrow -- but it's nearly midnight and I think I'm already the last one still awake...

bangerth commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+      /**
+       * Table for interpolating FastScape surface velocities back to ASPECT.
+       */
+      std::array< unsigned int, dim > table_intervals;
```
```suggestion
      std::array<unsigned int, dim> table_intervals;
```

bangerth commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+      double left_flux;
+
+      // FastScape erosional parameters //
+      /**
+       * Drainage area exponent for the stream power law (m variable in FastScape surface equation).
+       */
+      double m;
+
+      /**
+       * Slope exponent for the steam power law (n variable in FastScape surface equation).
+       */
+      double n;
+
+      /**
+       * Slope exponent for multi-direction flow, where 0 is uniform, and 10 is steepest descent. (-1 varies with slope)
+       * (p variable in FastScape surface equation).
+       */
+      double p;
+
+      /**
+       * Bedrock deposition coefficient. Higher values deposit more sediment
+       * inside the domain. (G variable in FastScape surface equation).
+       */
+      double g;
```
Could we use more descriptive names for these variables? Perhaps as follows?
```suggestion
      /**
       * Drainage area exponent for the stream power law. ($m$ variable in FastScape surface equation.)
       */
      double drainage_area_exponent_m;

      /**
       * Slope exponent for the steam power law. ($n$ variable in FastScape surface equation.)
       */
      double slope_exponent_n;

      /**
       * Slope exponent for multi-direction flow, where 0 is uniform, and 10 is steepest descent. (-1 varies with slope.)
       * ($p$ variable in FastScape surface equation.)
       */
      double slope_exponent_p;

      /**
       * Bedrock deposition coefficient. Higher values deposit more sediment
       * inside the domain. ($G$ variable in FastScape surface equation.)
       */
      double bedrock_deposition_g;
```

bangerth commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+       */
+      double h_erosional_base;
+
+      /**
+       * Surface porosity for sand.
+       */
+      double p1;
+
+      /**
+       * Surface porosity for silt.
+       */
+      double p2;
+
+      /**
+       * Sands e-folding depth for exponential porosity law (m).
+       */
+      double z1;
+
+      /**
+       * Silts e-folding depth for exponential porosity law (m).
+       */
+      double z2;
+
+      /**
+       * Sand-silt ratio
+       */
+      double r;
+
+      /**
+       * Averaging depth/thickness for sand-silt equation (m).
+       */
+      double l;
```
same issue here. single-letter variable names are just very difficult to understand in code if they are not purely local counting indices.

bangerth commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+
+  namespace MeshDeformation
+  {
+    /**
+     * Define FastScape functions as C functions. Must use exact same function/variable name
+     * and type as used in FastScape. All function names must be made lowercase, and an
+     * underscore added at the end. Types must be defined as pointers, and sent to
+     * FastScape as a reference. Additional functions are available within FastScape,
+     * see https://fastscape.org/fastscapelib-fortran/ for a list of all functions and
+     * their input parameters.
+     */
+#ifdef __cplusplus
+    extern"C" {
+#endif
```
This isn't strictly wrong, but I don't think you actually need these function declarations in the header file, do you? If you only need these functions in the `.cc` file, then this whole block of declarations might as well move to the `.cc` file.

bangerth reviewed (commented):
Here are my comments on the `.h` file. The `.cc` file is coming up next.

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+              in >> h[line];
+              line++;
+            }
+
+          in.close();
```
No need -- the file is closed where `in` goes out of scope.

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+      // Load in h values.
+      std::ifstream in;
+      in.open(restart_filename.c_str());
+      if (in)
+        {
+          unsigned int line = 0;
+
+          while (line < array_size)
+            {
+              in >> h[line];
+              line++;
+            }
+
+          in.close();
+        }
+
```
What do you do if `in` can not be read from? I think you want
```
  AssertThrow (in, ExcIO());
```
before this block, and then you can skip the `if (in)`.

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+        }
+
+      // Load in b values.
+      std::ifstream in_b;
+      in_b.open(restart_filename_basement.c_str());
+      if (in_b)
+        {
+          unsigned int line = 0;
+
+          while (line < array_size)
+            {
+              in_b >> b[line];
+              line++;
+            }
+
+          in_b.close();
+        }
```
Same here:
```suggestion
      std::ifstream in_b(restart_filename_basement);
      AssertThrow (in_b, ExcIO());
        {
          unsigned int line = 0;

          while (line < array_size)
            {
              in_b >> b[line];
              line++;
            }
          }
```

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+
+      // Load in silt_fraction values if
+      // marine sediment transport and deposition is active.
+      std::ifstream in_sf;
+      in_sf.open(restart_filename_silt_fraction.c_str());
+      if (use_marine_component && in_sf)
+        {
+          if (p1 > 0. || p2 > 0.)
+            this->get_pcout() << "   Restarting runs with nonzero porosity can lead to a different system after restart. " << std::endl;
+
+          unsigned int line = 0;
+
+          while (line < array_size)
+            {
+              in_sf >> sf[line];
+              line++;
+            }
+
+          in_sf.close();
+        }
```
```suggestion
      if (use_marine_component)
        {
          std::ifstream in_sf(restart_filename_silt_fraction);
          AssertThrow (in_sf, ExcIO());
          
          if (p1 > 0. || p2 > 0.)
            this->get_pcout() << "   Restarting runs with nonzero porosity can lead to a different system after restart. " << std::endl;

          unsigned int line = 0;

          while (line < array_size)
            {
              in_sf >> sf[line];
              line++;
            }
        }
```

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+      // Now load the FastScape istep at time of restart.
+      // Reinitializing FastScape always resets this to 0, so here
+      // we keep it in a separate variable to keep track for visualization files.
+      std::ifstream in_last_ouput_time;
+      in_last_ouput_time.open(restart_filename_time.c_str());
+      if (in_last_ouput_time)
+        {
+          in_last_ouput_time >> last_output_time;
+          in_last_ouput_time.close();
+        }
+    }
```
```suggestion
      std::ifstream in_last_ouput_time(restart_filename_time);
      AssertThrow (in_last_ouput_time, ExcIO());
        {
          in_last_ouput_time >> last_output_time;
        }
    }
```

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+      const std::string restart_filename_silt_fraction = dirname + "fastscape_sf_restart.txt";
+      const std::string restart_filename_time = dirname + "fastscape_last_output_time.txt";
+
+      std::ofstream out_h(restart_filename.c_str());
+      std::ofstream out_b(restart_filename_basement.c_str());
+      std::ofstream out_sf(restart_filename_silt_fraction.c_str());
+      std::ofstream out_last_output_time(restart_filename_time.c_str());
```
```suggestion
      std::ofstream out_h(restart_filename);
      std::ofstream out_b(restart_filename_basement);
      std::ofstream out_sf(restart_filename_silt_fraction);
      std::ofstream out_last_output_time(restart_filename_time);
```

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+      {
+        prm.enter_subsection ("Fastscape");
+        {
+          prm.declare_entry("Number of steps", "10",
```
See my comments earlier on on names for variables.

bangerth reviewed (commented):
@Djneu Here's the rest. I recognize that that is a lot of comments -- I apologize for the amount of work this is likely going to mean. Two notes on this:

* I've given hints how to do many of the places where you do MPI communication in a better way using the deal.II functions in namespace `Utilities::MPI`, but perhaps this is not necessary to implement right away: Your model works, and it isn't a specific impediment. If you don't want to redo these MPI things, then we can leave them for a follow-up patch.
* I've tried -- though not always succeeded -- to make as many comments into github "suggestions" which you can just commit. The indentation is likely wrong, but at least you don't have to copy-paste these changes into your source files. Once committed, you just have to make sure locally that it compiles, and run the indentation script. At the end, you just squash/fixup all of these individual commits.

bangerth commented:
Separately, I think that this PR is currently based on some commit from 2022. Can you rebase to current `main`?

anne-glerum commented:
@Djneu I pushed some documentation changes to your branch, this means you'll have to pull before you can push again.

Djneu commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+       */
+      double maximum_fastscape_timestep;
+
+      /**
+       * Check whether FastScape needs to be restarted. This is used as
+       * a mutable bool because we determine whether the model is being resumed in
+       * initialize(), and then after reinitializing FastScape we change it to zero
+       * so it does not initialize FastScape again in future timesteps.
+       * TODO: There is probably a better way to do this, and restarts should be rolled into
+       * the general ASPECT restart.
+       */
+      mutable bool restart;
```
I reset it to false during the fastscape_initialize function within the overarching read only function:

`    FastScape<dim>::compute_velocity_constraints_on_boundary(const DoFHandler<dim> &mesh_deformation_dof_handler,
                                                             AffineConstraints<double> &mesh_velocity_constraints,
                                                             const std::set<types::boundary_id> &boundary_ids) const`
                                                             
I've tried compiling it as a regular bool, but I get the following error:

`/cluster/project/sgtethz/shared_software/aspect/dneuharth/fastscape/aspect/source/mesh_deformation/fastscape.cc:365:27: error: assignment of member 'aspect::MeshDeformation::FastScape<2>::restart' in read-only object
  365 |                   restart = false;
`
Is there a way around this or is it fine to leave it as mutable?

Djneu commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+       */
+      mutable bool restart;
+
+      /**
+       * FastScape resets to step 0 on restart, this keeps the step number from the previous run.
+       * As we only want to read this value in from the restart file once, we initialize it in
+       * initialize() and only change it during the first timestep where FastScape is initialized.
+       */
+      mutable int restart_step;
```
This was actually an outdated variable, I've removed it now.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+    }
+
+
+    template <int dim>
+    void
+    FastScape<dim>::compute_velocity_constraints_on_boundary(const DoFHandler<dim> &mesh_deformation_dof_handler,
+                                                             AffineConstraints<double> &mesh_velocity_constraints,
+                                                             const std::set<types::boundary_id> &boundary_ids) const
+    {
+      if (this->get_timestep_number() == 0)
+        return;
```
Yes, if I remember correctly it was something like this. It looks like the mesh displacement system is one of the first things called, even before adding initial topography. Also, since there is no timestep length to yet run fastscape I figured it would be easiest to initialize and run it on timestep 1.

bangerth replied:
I see. Would you mind encoding your understanding of why you skip the first time step in a comment above these two lines?

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+        if (current_timestep == 1)
+          {
+            this->get_pcout() << "      Writing initial VTK..." << std::endl;
+            // Note: Here, the HHHHH field that is set to show the diffusivity (kd).
+            // However, you can change this so that any parameter is visualized.
+            fastscape_named_vtk_(kd.data(),
+                                 &vexp,
+                                 &visualization_step,
+                                 c,
+                                 &length);
```
I've updated the comment to be more descriptive. The original fastscape_vtk function would output only one field called HHHHH, which would show whichever data you put into it. We updated the function to by default output quite a few different fields, but not all of them, so I had left the HHHHH field as is in case someone might want to visualize something else, but by default it is sent the bedrock diffusivity data.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+    template <int dim>
+    bool FastScape<dim>::is_ghost_node(const unsigned int &index) const
+    {
+      int row = index / nx; // Calculate the row index
+      int col = index % nx; // Calculate the column index
+
+      // If we are at a boundary node and ghost nodes are enabled
+      // then return true.
+      if (row == 0 || row == ny-1 || col == 0 || col == nx-1)
+        {
+          if (use_ghost_nodes)
+            return true;
+          else
+            return false;
+        }
+      else
+        return false;
```
Yes, this seems like the better way to do it, thanks!

Djneu commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+       */
+      double end_time;
+
+      /**
+       * FastScape cell size in X, dx should always equal dy.
+       */
+      double dx;
+
+      /**
+       * FastScape cell size in Y, dy should always equal dx.
+       */
+      double dy;
```
I just double checked this, and these comments are actually not true and FastScape will still work if dx and dy are different, so it should work if not entirely square ASPECT cells are used with the coupling (though I do not know how not having square mesh elements affects FastScape results.)

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+        }
+      else
+        {
+          for (unsigned int i=0; i<temporary_variables.size(); ++i)
+            MPI_Ssend(&temporary_variables[i][0], temporary_variables[1].size(), MPI_DOUBLE, 0, 42, this->get_mpi_communicator());
```
I was having trouble calling isend this way, and it looks like isend/irecv are part of dealii 9.5 but not 9.4. Would it be best to leave it as is for now, and when ASPECT is updated to require dealii 9.5, I can update it to use these functions instead?

bangerth replied:
Ah, I didn't remember that we only added them for 9.5. In that case, yes, leave it as is with the explicit calls to MPI.

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+        }
+
+      //Functions::InterpolatedUniformGridData<dim> *velocities;
+      Functions::InterpolatedUniformGridData<dim> velocities (interpolation_extent,
+                                                              table_intervals,
+                                                              velocity_table);
+
+      VectorFunctionFromScalarFunctionObject<dim> vector_function_object(
+        [&](const Point<dim> &p) -> double
+      {
+        return velocities.value(p);
+      },
+      dim-1,
+      dim);
```
Thanks to @gassmoeller for helping improve this part!

Djneu commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+          // Find the appropriate sediment rain based off the time interval.
+          const double time_in_years = this->get_time() / year_in_seconds;
+          double sediment_rain = sediment_rain_rates[0];
+          for (unsigned int j=0; j<sediment_rain_times.size(); ++j)
+            {
+              if (time_in_years > sediment_rain_times[j])
+                sediment_rain = sediment_rain_rates[j+1];
+            }
```
I've added an assert to make sure it is increasing. What would be a good way to shortcut this?

bangerth replied:
The assertion should go into the function that reads the parameters, so that the error is shown as early as possible if a user uses non-ascending times.

To make this code shorter, you might want to use `std::lower_bound()`, which uses a binary search through the array.

Djneu commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+       */
+      // mutable bool fastscape_mesh_filled = true;
+
+      /**
+       * ASPECT end time to check if we should destroy FastScape.
+       */
+      double end_time;
```
Just to double check, I couldn't find a specific function to get the end_time from simulator access. Is there one somewhere, or should I write a new one?

bangerth replied:
Yes, just add the corresponding function to `SimulatorAccess`.

Djneu replied:
Okay, I've changed it so that end_time is read in in parameters.cc, and then this can be accessed through a get_end_time SimulatorAccess function.

bangerth commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,725 @@
+                              const double *vexp,
+                              int *astep,
+                              const char *c,
+                              int *length);
```
Strictly speaking, you can put anything you want here and the compiler will not complain -- that's because the compiler will simply take your word that that is the signature of the function, and it's up to *you* to make sure it matches what the Fortran function *really* does.

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1739 @@
+    template <int dim>
+    FastScape<dim>::~FastScape ()
+    {
+      // It doesn't seem to matter if this is done on all processors or only on the one that runs
+      // FastScape as the destroy function checks if the memory is allocated.
+      if (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0)
+        fastscape_destroy_();
```
I think what confused me is that `fastscape_init_()` is called in `initialize_fastscape()`, which does not have the check for `if (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0)` in it. So it *looks* like fastscape is initialized on all processes. But that function is only called on process zero.

Perhaps you can make that clear by putting the following into `fastscape_initialize()`, at the top:
```
  Assert (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0, ExcInternalError());
```

bangerth commented:
Thanks, @Djneu , for addressing my comments. Can you push your code changes? I've got a couple more questions where I wanted to look at how the new code looks that you mention in your replies.

When pushing, make sure you don't overwrite @anne-glerum 's commits.

Djneu commented:
Hi @bangerth, thank you for all the comments! I've updated my branch to where I've gotten to so far, which should be most of the changes in fastscape.cc and fastscape.h that were mentioned. 

I've also added a python script in the contrib folder, this is something that is necessary to visualize ASPECT and FastScape visualizations at the same time, as the general FastScape vtk files do not write a time. The script converts the VTK files into VTS using the solution.pvd times so they will match up.

Djneu commented:
Hi all,
Thanks again for all the suggestions! Other than a few minor comments (e.g., how to handle the GFORTRAN_CONVERT_UNIT), I think I've addressed all the comments and the code is ready for another review. @bangerth @gassmoeller @anne-glerum @naliboff @tjhei

bangerth reviewed (approved):
We've been talking about it at a ASPECT get-together yesterday and today and decided that this is good enough for now and that we should merge. If we find other things that need to be addressed, we can always do that in follow-ups, but it doesn't make it easier by carrying around a 3000-line patch.

From my side, my apologies for letting this sit for another 3 months after the hackathon!

bangerth commented:
But it turns out that I can't merge, even though I want, github just doesn't let me :-) You'll have to address the following things:
* Run the `./contrib/utilities/indent` script.
* There's a test that fails and that looks like this:
```
 302/1045 Test  #302: free_surface_VE_cylinder_2D_loading_fixed_elastic_dt_gmg ...........***Failed    0.66 sec
*** "Test free_surface_VE_cylinder_2D_loading_fixed_elastic_dt_gmg failed": ***
[1/1] Generating output-free_surface_VE_cylinder_2D_loading_fixed_elastic_dt_gmg/screen-output
FAILED: output-free_surface_VE_cylinder_2D_loading_fixed_elastic_dt_gmg/screen-output 
cd /__w/aspect/aspect/build/tests && /__w/aspect/aspect/tests/cmake/timeout.sh 600s mpirun -np 2 /__w/aspect/aspect/build/aspect /__w/aspect/aspect/build/tests/free_surface_VE_cylinder_2D_loading_fixed_elastic_dt_gmg.x.prm > /__w/aspect/aspect/build/tests/output-free_surface_VE_cylinder_2D_loading_fixed_elastic_dt_gmg/screen-output.tmp && /__w/aspect/aspect/tests/cmake/default screen-output </__w/aspect/aspect/build/tests/output-free_surface_VE_cylinder_2D_loading_fixed_elastic_dt_gmg/screen-output.tmp >/__w/aspect/aspect/build/tests/output-free_surface_VE_cylinder_2D_loading_fixed_elastic_dt_gmg/screen-output

--------------------------------------------------------
An error occurred in line <1565> of file </opt/tmp/unpack/deal.II-master/include/deal.II/lac/la_parallel_vector.h> in function
    Number dealii::LinearAlgebra::distributed::Vector<Number, MemorySpace>::operator()(dealii::LinearAlgebra::distributed::Vector<Number, MemorySpace>::size_type) const [with Number = double; MemorySpace = dealii::MemorySpace::Host; dealii::LinearAlgebra::distributed::Vector<Number, MemorySpace>::size_type = unsigned int]
The violated condition was: 
    partitioner->in_local_range(global_index) || vector_is_ghosted == true
Additional information: 
    You tried to read a ghost element of this vector, but it has not
    imported its ghost values.
```
Do you have an idea why that might fail? Several other tests fail with the same error. It is possible that if you rebase to the current `main` branch, the error will go away. Can you try a rebase?

* While you're there, and doing the rebase anyway, would you mind merging the commits into one (or one for the changes, one for tests, one for the changelog entry, one for documentation, etc -- i.e., into separate, self-contained commits)?

Thanks for your patience with this!

gassmoeller commented:
/rebuild

Djneu commented:
Hi all,
It looks like both the failed checks are related to the fastscape_eroding_box cookbook. It makes sense this won't run as fastscape.cc isn't compiled. Is there anything I should do for this?

gassmoeller commented:
Ah, I forgot about this, yes, there are two simple fixes:
1. The cookbook markdown file was not listed in the cookbook index, and it didnt contain a proper markdown heading, this caused warnings in our documentation tester.
2. We automatically test if all cookbooks prms at least conform to proper .prm file syntax. Since fastscape is not compiled by default the cookbook you added had to manually be listed as exempt from that test.

I pushed a commit that should fix both issues. Let's see what the testers say, if they pass, I will merge.
